src/Symfony/Component/Scheduler/Trigger/CronExpressionTrigger.php

```diff
@@ -42,8 +63,42 @@ public static function fromSpec(string $expression = '* * * * *'): self
         return new self(new CronExpression($expression));
     }
 
+    public static function fromHash(string $expression, string $context): self
+    {
+        return self::fromSpec(self::parseHashed($expression, $context));
+    }
+
     public function getNextRunDate(\DateTimeImmutable $run): ?\DateTimeImmutable
     {
         return \DateTimeImmutable::createFromMutable($this->expression->getNextRunDate($run));
     }
+
+    private static function parseHashed(string $expression, string $context): string
+    {
+        $expression = self::HASH_ALIAS_MAP[$expression] ?? $expression;
+        $parts = explode(' ', $expression);
+
+        if (5 !== \count($parts)) {
+            return $expression;
+        }
+
+        foreach ($parts as $position => $part) {
+            if (preg_match('#^\#(\((\d+)-(\d+)\))?$#', $part, $matches)) {
+                $parts[$position] = self::hashField(
+                    (int) ($matches[2] ?? self::HASH_RANGES[$position][0]),
+                    (int) ($matches[3] ?? self::HASH_RANGES[$position][1]),
+                    $context
+                );
+            }
+        }
+
+        return implode(' ', $parts);
+    }
+
+    private static function hashField(int $start, int $end, string $context): int
+    {
+        $possibleValues = range($start, $end);
+
+        return $possibleValues[(int) fmod(hexdec(substr(md5($context), 0, 10)), \count($possibleValues))];
```

Comment:
This algorithm creates two biases:

1. It does not take the field into account, thus the fields will not be randomized independently.
2. If the range does not cleanly divide `1099511627775` (the amount of different values that may be returned by the hexdec), lower values in the range will be preferred.

With PHP 8.2, the `\Random\Randomizer` combined with a seeded engine will solve both of these issues. See also example 2 here: https://www.php.net/manual/en/random-engine-xoshiro256starstar.construct.php

Reply:
Interesting, I borrowed/modified this code from [`UrlPackage::chooseBaseUrl()`](https://github.com/symfony/symfony/blob/5b4236a104979b402a547dde8d46697aa5501e00/src/Symfony/Component/Asset/UrlPackage.php#L110). That code was actually [updated recently](https://github.com/symfony/symfony/blob/847416f5711f53867236d6a9c4933b7da2f12605/src/Symfony/Component/Asset/UrlPackage.php#L110). Does the new version alleviate either of the described biases?

Reply:
To my understanding (1) does not apply there, because there are no related values (a full URL is randomized). To clarify the issue (1), consider that you randomize all fields with the range 1-5 (for whatever reasons), then all the chosen values will be identical, unless I missed anything.

(2) also affects the linked function. Anytime a remainder is calculated when the range of the dividend is not a multiple of the divisor, a bias will be introduced. Let me give a simple example, consider the randomly chosen value is in the range from 0-15 and you want to map that to the range from 0-5 (6 possible values), then 0-11 will cleanly map onto 0-5, and 12, 13, 14, 15 will map to 0, 1, 2, 3 and thus chosen with a slightly higher probability. To avoid this bias, rejection sampling needs to be used: Instead of reducing the value by calculating the remainder, a new random value is obtained. This can be optimized by only “rejecting” the values that are greater than the largest multiple of the divisor. 

In the specific case of `chooseBaseUrl()`, selection performance is likely more important than a completely uniform selection, especially since the bias is usually small. That's likely also why `crc32` is used, as that hash function is pretty fast. However in the case of cronjobs, selection performance does not really matter and you'd rather want to spread out the jobs as uniformly as possible.

The `\Random\Randomizer`, like `\random_int()` implements proper rejection sampling to ensure a completely unbiased result (given an unbiased engine).

Reply:
> consider that you randomize all fields with the range 1-5 (for whatever reasons), then all the chosen values will be identical, unless I missed anything.

Ok, yeah I understand.

> The \Random\Randomizer, like \random_int() implements proper rejection sampling to ensure a completely unbiased result (given an unbiased engine).

Could you help with developing the required code? You clearly understand this stuff much better than me! :)

Reply:
> You clearly understand this stuff much better than me! :)

I contributed quite a few design decisions for the ext/random RFC and am now one of the maintainers of it. I can't leave out an opportunity to advertise the new [randomness interface](https://www.php.net/manual/en/book.random.php) of PHP 8.2 :smile: I found this PR by pure chance, via your issue in dragonmantank/cron-expression, because I wanted to build this exact feature on top of that library for our non-Symfony application at work.

-----------------------------

Let me give you the PHP 8.2+ solution first, to showcase it:

```php
$r = new \Random\Randomizer(new \Random\Engine\Xoshiro256StarStar(\hash('sha256', $context, true)));

// now for each field:
$r->getInt($min, $max);
```

-----------------------------

For versions before PHP 8.2, the following should do the trick (taken from PHP 8.3's `getBytesFromString()`):

1. Create a generator that emits one byte each for a given context. More specifically the numeric byte value:

```php
/** Returns a uniform stream of numbers from 0 to 255 seeded by the given context. */
function byteStream(string $context): Iterator {
  $hash = $context;
  while (true) {
    $hash = \hash('sha256', $hash, true);
    for ($i = 0; $i < \strlen($hash); $i++) {
      yield ord($hash[$i]);
    }
  }
}
```

2. Given the generator, generate random values for the fields. By definition the maximum range is 60 (minute field), thus a single byte provides sufficient randomness.

```php
$byteStream = byteStream($context);
$range = $max - $min;
\assert($range <= 0xff);
$mask = $range;      // This masks of the higher bits that will
$mask |= $mask >> 1; // never be used in order to avoid unnecessarily
$mask |= $mask >> 2; // rejecting values.
$mask |= $mask >> 4; // https://github.com/php/php-src/pull/10522
do {
 $random = $byteStream->current() & $mask; // apply the mask
 $byteStream->next();
} while ($random > $range); // reject the value if the range is exceeded.

$value = $random + $min;
```

Example: https://3v4l.org/fLCNM

Consider this freely licensed for inclusion in Symfony and dragonmantank/cron-expression. Use `Co-authored-by: Tim Düsterhus <tim@bastelstu.be>` if you'd like.

Reply:
> I contributed quite a few design decisions for the ext/random RFC and am now one of the maintainers of it. I can't leave out an opportunity to advertise the new [randomness interface](https://www.php.net/manual/en/book.random.php) of PHP 8.2

That's awesome, thanks for chipping in here!

> Let me give you the PHP 8.2+ solution first, to showcase it:

This looks great! What a simplification of the below code!

> Consider this freely licensed for inclusion in Symfony and dragonmantank/cron-expression. Use Co-authored-by: Tim Düsterhus <tim@bastelstu.be> if you'd like.

Thanks! I'll see what the other reviewers would like to do here. As Fabien said above, it would be idea if this lived in dragonmantank/cron-expression. Then both Symfony and you could benefit!

For the record, I've been using my _naive_ function on dozens of cron jobs for years without issue (it has them all run at different times consistently which is what I was after at the end of the day). _I absolutely_ want to add the 8.2+ method but not sure of the _polyfill_...

Thanks again Tim!

Reply:
> it has them all run at different times consistently which is what I was after at the end of the day

Yeah, the bias is small, but depending on the use-case it can be meaningful.

The use-case I'm after for our application is randomizing the time of the cronjob for the automated updater by seeding the times depending on the UUID identifying the instance, to ensure the update requests are evenly distributed during the day. However 1440 minutes in a day does not cleanly divide any power of two, thus there will be a bias and certain times are more likely to be selected, thus unevenly distributing the load on the update servers. It's fine for us, because we have plenty of spare capacity (and older versions used a fixed date anyway :see_no_evil:), but for Symfony users with similar requirements regarding evenly distributing the load it could possibly make a significant difference in server costs (due to autoscaling etc).

> Thanks! I'll see what the other reviewers would like to do here.

I'll leave this decision up to you, as I'm no Symfony user. I agree though, that it would be great to have this upstreamed in dragonmantank/cron-expression.

> but not sure of the polyfill...

Functionally the https://github.com/arokettu/php-random-polyfill polyfill is great, but it necessarily is pretty slow/heavy. Emulating random engines without unsigned math is fairly expensive.

Reply:
I would keep things simple by keeping the current implementation for PHP < 8.2, and using the native Randomizer when possible.

Reply:
It looks like you forgot to use the native Randomizer class for PHP >= 8.2

Reply:
Ok, I misunderstood what you meant. I'll add.

Reply:
@kbond Just to have it noted: Currently the fields will still not be randomized independently (i.e. `#(1, 2)` will always result in the same value for all fields of a single cron expression). 

Incidentally I've implemented the same feature just a few days ago for our application, opting for this solution: https://github.com/WoltLab/WCF/blob/3a5cc20c98d35728b29149ad983655d26fa1befa/wcfsetup/install/files/lib/system/package/plugin/CronjobPackageInstallationPlugin.class.php#L112-L134

For the non-PHP 8.2 version, the simplest solution would be adding a `counter` to the md5 value, something like:

```
$counter = 0;
$engine = function ($start, $end) use ($context, &$counter) {
    $possibleValues = range($start, $end);
    $counter++;

    return $possibleValues[(int) fmod(hexdec(substr(md5($context . '-' . $counter), 0, 10)), \count($possibleValues))];
};
```

And then generating each field as `$engine($min, $max);` and with 8.2:

```
$randomizer = new Randomizer(new Xoshiro256StarStar(hash('sha256', $context, true)));
$engine = fn ($start, $end) => $randomizer->getInt($start, $end);
```

Reply:
> Currently the fields will still not be randomized independently (i.e. #(1, 2) will always result in the same value for all fields of a single cron expression).

To clarify, this is a problem on our 8.2+ code?

Reply:
> To clarify, this is a problem on our 8.2+ code?

Yes, you are creating a new instance of the xoshiro256** engine for each call to `hashField()` and the `$context` (which acts as the seed) is the same for all fields in line 95:

https://github.com/symfony/symfony/pull/49792/files#diff-79c7bf265e6bc52bb6ccb0609ff997b5f3414c06cd059fe37c79cf2904825c00R95

Reply:
Got it, will fix that up. Thanks!

Reply:
@TimWolla, currently trying to implement your above code. For the same `$context`, the hash `# # # * *` and `# #(0-2) # * *` (both PHP 8.2+ and <8.2 versions). Is this expected or did I mess something up?

Reply:
> Is this expected

Perhaps! It might be possible that your context seed causes `#(0-2)` and `#(0-23)` (if that's what `#` does for the hour) results in the same value.

Basically you would need the following test case:

```
#(1-6) #(1-6) #(1-6) #(1-6) #(1-6)
```

If the values are different, it works correctly, because then each field is independently generated. If some values are the same, it's fine, that's just how randomness works.

Of course with a different `$context` the result should be different, but again, if some values are the same, that's just how randomness works.